### PR TITLE
issue-130 copy contents related x.py from rust-forge to rustc-guide

### PR DESCRIPTION
**Pull request by rajcspsg**

### CODE REVIEW

File: src/compiler-documenting.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+   ./x.py doc src/doc/book
+   ./x.py doc src/doc/nomicon
+   ./x.py doc src/doc/book src/libstd
+```
+
+Much like individual tests or building certain components you can build only
+ the documentation you want.
+
+## Document internal rustc items
+By default rustc does not build the compiler docs for its internal items.
```
Comment (mark-i-m) [inline src/compiler-documenting.md]: @davidtwco Is this still true? How is the [rustc rustdoc](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/) built?

File: src/compiler-documenting.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+By default rustc does not build the compiler docs for its internal items. 
+Mostly because this is useless for the average user. However, you might need 
+to have it available so you can understand the types. Here’s how you can 
+compile it yourself. From the top level directory where x.py is located run:
+
+    cp config.toml.example config.toml
+
+Next open up config.toml and make sure these two lines are set to true:
+
+docs = true
```
Comment (mark-i-m) [inline src/compiler-documenting.md]: These two lines should be inside ` ``` `?

File: src/compiler-documenting.md (hunk 1 above)
Comment (davidtwco) [reply src/compiler-documenting.md]: Compiler documentation is not built by default - there's a `config.toml` flag for that. But, when enabled, compiler documentation does [include internal items](https://github.com/rust-lang/rust/blob/79fcc58b24d85743d025fd880fca55748662ed3e/src/bootstrap/doc.rs#L699).

File: src/build-install-distribution-artifacts.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+```
+
+
+# Install distribution artifacts
+
+If you’ve built a distribution artifact you might want to install it and 
+test that it works on your target system. You’ll want to run this command:
+
+   `./x.py install`
+
```
[the remark this answers is not in the document]
Comment (rajcspsg) [reply src/build-install-distribution-artifacts.md]: @mark-i-m I didn't get this comment. could you please explain the note that needs to be added here.

File: src/compiler-benchmarking.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Benchmarking rustc
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply src/compiler-benchmarking.md]: I hadn't heard of these. I just tried `./x.py bench`, it looks like it's basically the same as `cargo bench`, i.e. it runs `#[bench]` unit benchmarks. That might be useful for benchmarking certain components within the compiler (mostly libraries in `std`) but it doesn't give any info on the overall speed of the compiler. Also, I got compile errors for a bunch of these, so it looks like they aren't being run regularly.

**Comment by rajcspsg** [comment]
@mark-i-m please let me know your thoughts on pending items I mentioned [above](https://github.com/rust-lang-nursery/rustc-guide/pull/195#issuecomment-422573669).

File: src/how-to-build-and-run.md [hunk 5]
```diff
@@ -47,6 +47,14 @@ debuginfo-lines = true
 use-jemalloc = false
 ```
 
+### what is x.py?
```
Comment (mark-i-m) [inline src/how-to-build-and-run.md]: Nit: "what" -> "What"

File: src/how-to-build-and-run.md [hunk 6]
```diff
@@ -80,6 +88,52 @@ compiling `rustc` is done in stages:
...
+        --target TARGET target targets to build
+        --on-fail CMD   command to run on failure
+        --stage N       stage to build
+        --keep-stage N  stage to keep without recompiling
+        --src DIR       path to the root of the rust checkout
+    -j, --jobs JOBS     number of jobs to run in parallel
+    -h, --help          print this help message
+```
+
+One thing to keep in mind is that `rustc` is a _bootstrapping_ compiler. That
```
Comment (mark-i-m) [inline src/how-to-build-and-run.md]: This is already discussed in the previous section. I think we don't need lines added 114-136.

File: src/tests/running.md [hunk 7]
```diff
@@ -93,3 +93,24 @@ This is much faster, but doesn't always work. For example, some tests
 include directives that specify specific compiler flags, or which rely
 on other crates, and they may not run the same without those options.
 
+### Run only the tidy script
```
Comment (mark-i-m) [inline src/tests/running.md]: Perhaps move these to the end of the "Running a subset of the test suites" section above (in the same file).

File: src/build-install-distribution-artifacts.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# Build distribution artifacts
+
+You might want to build and package up the compiler for distribution.
+You’ll want to run this command to do it:
+
+   `./x.py dist`
+
+Other Flags
```
[the remark this answers is not in the document]
Comment (mark-i-m) [reply src/build-install-distribution-artifacts.md]: Yes, I think we can remove them from both places. The flags are covered in a lot more detail elsewhere.

File: src/build-install-distribution-artifacts.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mark-i-m) [reply src/build-install-distribution-artifacts.md]: If you are testing out a modification to a compiler, you might want to use it to compile some project. Usually, you _do not_ want to use `./x.py install` for testing. Rather, you should create a toolchain as discussed in how-to-build-and-run.html#creating-a-rustup-toolchain. For example, if the toolchain you created is called `foo`, you would then invoke it with `rustc +foo ...` (where ... represents the rest of the arguments).

If you want you can just copy paste what I wrote.

**Comment by mark-i-m** [comment]
@rajcspsg Thanks! Sorry for the delay. I have been a bit swamped lately... I left some comments (and I think I answered your questions; let me know if I haven't). Overall, this is looking good :)

File: src/build-install-distribution-artifacts.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rajcspsg) [reply src/build-install-distribution-artifacts.md]: This is done.

File: src/compiler-benchmarking.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rajcspsg) [reply src/compiler-benchmarking.md]: this file is removed.

File: src/compiler-documenting.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# Documenting rustc
+
+You might want to build documentation of the various components 
+available like the standard library. There’s two ways to go about this.
+ You can run rustdoc directly on the file to make sure the HTML is 
+ correct which is fast or you can build the documentation as part of the
```
[the remark this answers is not in the document]
Comment (rajcspsg) [reply src/compiler-documenting.md]: this is done.

File: src/how-to-build-and-run.md (hunk 5 above)
Comment (rajcspsg) [reply src/how-to-build-and-run.md]: This is done.

File: src/how-to-build-and-run.md (hunk 6 above)
Comment (rajcspsg) [reply src/how-to-build-and-run.md]: This is done.

File: src/tests/running.md (hunk 7 above)
Comment (rajcspsg) [reply src/tests/running.md]: This is done.

**Comment by rajcspsg** [comment]
@mark-i-m I have updated with your feedback. Please check and let me know if any changes are required.

File: src/build-install-distribution-artifacts.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+If you’ve built a distribution artifact you might want to install it and
+test that it works on your target system. You’ll want to run this command:
+
+   ```bash
+   ./x.py install
+   ```
+
+   Note: If you are testing out a modification to a compiler, you might want to use it to compile some project.
+   Usually, you do not want to use ./x.py install for testing.
+   Rather, you should create a toolchain as discussed in how-to-build-and-run.html#creating-a-rustup-toolchain.
```
Comment (mark-i-m) [inline src/build-install-distribution-artifacts.md]: Could you make this a link to the actual chapter please?

File: src/build-install-distribution-artifacts.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+   For example, if the toolchain you created is called foo, you would then invoke it with rustc +foo ... (where ... represents the rest of the arguments).
```
Comment (mark-i-m) [inline src/build-install-distribution-artifacts.md]: Could you put the `rustc +foo ...` part in backticks so it formats as monospace font?

**Comment by mark-i-m** [comment]
@rajcspsg Looking good! To minor nits, then r=me :+1: 

Nice job :)

File: src/build-install-distribution-artifacts.md [hunk 12]
```diff
@@ -16,7 +16,14 @@ test that it works on your target system. You’ll want to run this command:
...
-   For example, if the toolchain you created is called foo, you would then invoke it with rustc +foo ... (where ... represents the rest of the arguments).
\ No newline at end of file
+   Rather, you should create a toolchain as discussed in below link.
+    [creating-a-rustup-toolchain][create-rustup-toolchain]
+
+   For example, if the toolchain you created is called foo, you
+   would then invoke it with `rustc +foo ...` (where ... represents
+   the rest of the arguments).
+
+[create-rustup-toolchain]: https://github.com/rust-lang-nursery/rustc-guide/blob/master/src/how-to-build-and-run.md#creating-a-rustup-toolchain
```
Comment (mark-i-m) [inline src/build-install-distribution-artifacts.md]: Sorry, I should have been more clear. This should be `./how-to-build-and-run.md#creating-a-rustup-toolchain`

Comment (rajcspsg) [reply src/build-install-distribution-artifacts.md]: [create-rustup-toolchain]: ./how-to-build-and-run.md#creating-a-rustup-toolchain

Does the above one work?

**Comment by mark-i-m** [comment]
:+1: Thanks!

**Comment by rajcspsg** [comment]
@mark-i-m does this PR covers the 4th point in this [issue](https://github.com/rust-lang-nursery/rustc-guide/issues/100)?

**Pull request merged**